system/jlib/jencrypt.cpp

```diff
@@ -1829,19 +1833,195 @@ size_t aesDecrypt(const void *key, size_t keylen, const void *input, size_t inle
+        int ciphertext_len = 0;
+        int thislen = 0;
+        if(1 != EVP_EncryptUpdate(ctx, ciphertext, &thislen, (const unsigned char *) plaintext, plaintext_len))
+            encryptError("Error in EVP_EncryptUpdate");
+        ciphertext_len += thislen;
+        if(1 != EVP_EncryptFinal_ex(ctx, ciphertext + ciphertext_len, &thislen))
+            encryptError("Error in EVP_EncryptFinal_ex");
+        ciphertext_len += thislen;
+        EVP_CIPHER_CTX_free(ctx);
+        output.setLength(ciphertext_len);
+        return output;
+    }
+    catch (...)
+    {
+        EVP_CIPHER_CTX_free(ctx);
+        throw;
+    }
+}
+
+static void decryptError(const char *what)
+{
+    VStringBuffer s("openssl::aesDecrypt: unexpected failure in %s", what);
+    throw makeStringException(0, s.str());
+}
+
+MemoryBuffer &aesDecrypt(const void *key, size_t keylen, const void *ciphertext, size_t ciphertext_len, MemoryBuffer &output)
+{
+    EVP_CIPHER_CTX *ctx;
+
+    int thislen = 0;
+    int plaintext_len = 0;
+
+    if(!(ctx = EVP_CIPHER_CTX_new()))
+        decryptError("Failed to create context");
+    try
+    {
+        unsigned char iv[16] = { 0 };
+        switch (keylen)
+        {
+            case 32:
+                if(1 != EVP_DecryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, (const unsigned char *) key, iv))
+                    decryptError("Failed to initialize context");
+                break;
+            case 24:
+                if(1 != EVP_DecryptInit_ex(ctx, EVP_aes_192_cbc(), NULL, (const unsigned char *) key, iv))
+                    decryptError("Failed to initialize context");
+                break;
+            case 16:
+                if(1 != EVP_DecryptInit_ex(ctx, EVP_aes_128_cbc(), NULL, (const unsigned char *) key, iv))
+                    decryptError("Failed to initialize context");
+                break;
+            default:
+                decryptError("Unsupported key length");
+                break;
+        }
+        byte *plaintext = (byte *) output.reserve(ciphertext_len + 100);
+        if(1 != EVP_DecryptUpdate(ctx, plaintext, &thislen, (const unsigned char *) ciphertext, ciphertext_len))
+            decryptError("Error in EVP_DecryptUpdate");
+        plaintext_len += thislen;
+
+        if(1 != EVP_DecryptFinal_ex(ctx, plaintext + plaintext_len, &thislen))
+            decryptError("Error in EVP_DecryptFinal_ex");
+        plaintext_len += thislen;
+        EVP_CIPHER_CTX_free(ctx);
+        return output;
+    }
+    catch (...)
+    {
+        EVP_CIPHER_CTX_free(ctx);
+        throw;
+    }
+}
+
+size_t aesDecrypt(const void *key, size_t keylen, const void *ciphertext, size_t ciphertext_len, void *output, size_t outlen)
+{
+    EVP_CIPHER_CTX *ctx;
+
+    int thislen = 0;
+    size_t plaintext_len = 0;
+
+    if(!(ctx = EVP_CIPHER_CTX_new()))
+        decryptError("Failed to create context");
+
+    try
+    {
+        if (outlen < ciphertext_len)
```

Comment:
I think this is too strict.  For instance I changed the unit test to the following, and it fails:

```
       unsigned char key[] = { 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
                                0x38, 0x39, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35,
                                0x36, 0x37, 0x38, 0x39, 0x30, 0x31, 0x32, 0x33,
                                0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x31
                              };

        /* Message to be encrypted */
        constexpr const char * text = "The quick brown fox jumps over the lazy dog";
        constexpr unsigned lenText = strlen(text);
        const unsigned char *plaintext = (const unsigned char *)text;

        MemoryBuffer ciphertext1, ciphertext2, decrypted;

        openssl::aesEncrypt(key, 32, plaintext, strlen ((char *)plaintext), ciphertext1);
        jlib::aesEncrypt(key, 32, plaintext, strlen ((char *)plaintext), ciphertext2);

        CPPUNIT_ASSERT(ciphertext1.length()==ciphertext2.length());
        CPPUNIT_ASSERT(memcmp(ciphertext1.bytes(), ciphertext2.bytes(), ciphertext1.length()) == 0);
        
        /* Decrypt the ciphertext */
        openssl::aesDecrypt(key, 32, ciphertext1.bytes(), ciphertext1.length(), decrypted);

        char decryptedBuffer[lenText+1];
        openssl::aesDecrypt(key, 32, ciphertext1.bytes(), ciphertext1.length(), decryptedBuffer, lenText);
        decryptedBuffer[lenText] = 0;

        /* Add a NULL terminator. We are expecting printable text */
        decrypted.append('\0');

        /* Show the decrypted text */
        DBGLOG("Decrypted text is: %s/%s", (const char *) decrypted.bytes(), decryptedBuffer);

```

Reply:
If we need to retain this test, then we need to check that everywhere that calls this function has allocated the correct amount of memory - with any last block padding necessary.